### PR TITLE
added doc for LDAP sync error tolerances

### PR DESCRIPTION
**Pull request by stevekuznetsov**
This PR documents the new error tolerance booleans provided in https://github.com/openshift/origin/pull/7263

@liggitt PTAL

@adellape FYI

### CODE REVIEW

File: install_config/syncing_groups_with_ldap.adoc [hunk 1]
```diff
@@ -329,6 +331,13 @@ upgraded to TLS.
 <5> The attribute on the group that stores the membership information.
 <6> The attribute that uniquely identifies a user on the LDAP server.
 <7> The attribute to use as the name of the user in the OpenShift Group record.
+<8> When `true`, the sync job will tolerate groups for which some members were not found.
+If this is `true`, members whose LDAP entries are not found will be ignored. The default
+behavior for the sync job is to fail if a member of a group is not found.
+<9> When `true`, the sync job will tolerate groups for which some members are outside
+the user scope given in the `usersQuery` base DN. If this is `true`, out-of-scope members
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: "outside the scope..."

File: install_config/syncing_groups_with_ldap.adoc [hunk 2]
```diff
@@ -318,6 +318,8 @@ rfc2307:
         filter: (objectclass=inetOrgPerson)
     userUIDAttribute: dn <6>
     userNameAttributes: [ mail ] <7>
+    tolerateMemberNotFoundErrors: true <8>
+    tolerateMemberOutOfScopeErrors: true <9>
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: add the attributes to all the rfc sample configs, but set to `false` elsewhere

**Comment by liggitt** [comment]
I wonder if it would be good to have a third rfc2307 config instead that puts the explanation of the error skipping in the main doc, rather than in the config footnotes (so "RFC2307", "RFC2307 with user-defined mappings", "RFC2307 with optional error settings"). I think it's important to explain why you only want to turn these on once you've verified the errors you're getting are ones you want to tolerate (and what happens if you ignore these errors with a misconfigured query... your group gets emptied)

**Comment by stevekuznetsov** [comment]
@liggitt PTAL

File: install_config/syncing_groups_with_ldap.adoc [hunk 3]
```diff
@@ -433,6 +433,153 @@ users:
...
+query is properly configured, these errors can be tolerated and the out of scope members
+ignored.
+
+Similarly, when the group sync process fails to locate a member for a group, it fails
+outright. Failures of this type most often indicate mis-configured LDAP group configurations
+on the LDAP server. However, if it is certain that the group entries on the LDAP server
+have been configured correctly, these errors can be tolerated and the problematic
+members ignored.
+
+[NOTE]
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: make this a warning

File: install_config/syncing_groups_with_ldap.adoc [hunk 4]
```diff
@@ -433,6 +433,153 @@ users:
...
+====
+Enabling error tolerances for the LDAP group sync will cause the sync process to ignore
+problematic member entries. If the LDAP group sync is not configured correctly, this will
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: s/will/could/

File: install_config/syncing_groups_with_ldap.adoc [hunk 5]
```diff
@@ -433,6 +433,153 @@ users:
 <1> The name of the Group as specified by the user-defined name mapping.
 ====
 
+[[rfc2307-with-error-tolerances]]
+=== RFC 2307 with User-Defined Error Tolerances
+
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: Probably too much detail. Just two paragraphs, one for each error type, include the actual error text (will help troubleshooting searches find this section), a short description of what causes it, and mention the config field name.

File: install_config/syncing_groups_with_ldap.adoc [hunk 6]
```diff
@@ -433,6 +433,153 @@ users:
...
+    groupMembershipAttributes: [ member ]
+    usersQuery:
+        baseDN: "ou=users,dc=example,dc=com"
+        scope: sub
+        derefAliases: never
+        filter: (objectclass=inetOrgPerson)
+    userUIDAttribute: dn
+    userNameAttributes: [ mail ]
+    tolerateMemberNotFoundErrors: true <1>
+    tolerateMemberOutOfScopeErrors: true <2>
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: include the config fields, set to `false` in all the rfc sample configs so when they copy/paste them, they'll have them in their configs

**Comment by stevekuznetsov** [comment]
@liggitt ready for another look

File: install_config/syncing_groups_with_ldap.adoc [hunk 7]
```diff
@@ -433,6 +437,155 @@ users:
...
+=== RFC 2307 with User-Defined Error Tolerances
+
+By default, if groups being synced contain members whose entries that are outside of the
+scope defined in the member query, the group sync fails with an error like:
+
+----
+Error determining LDAP group membership for "<group>": membership lookup for user "<user>" in group "<group>" failed because of "search for entry with dn="<user-dn>" would search outside of the base dn specified (dn="<base-dn>")".
+----
+
+Most often, failures of this type indicated a mis-configured user query. However, if it
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: This often indicates a mis-configured `baseDN` in the `usersQuery`. However, in cases where the `baseDN` intentionally does not contain some of the members of the group, setting `tolerateMemberOutOfScopeErrors: true` will allow the group sync to continue.

File: install_config/syncing_groups_with_ldap.adoc [hunk 8]
```diff
@@ -433,6 +437,155 @@ users:
...
+----
+Error determining LDAP group membership for "<group>": membership lookup for user "<user>" in group "<group>" failed because of "search for entry with base dn="<user-dn>" refers to a non-existent entry".
+
+Error determining LDAP group membership for "<group>": membership lookup for user "<user>" in group "<group>" failed because of "search for entry with base dn="<user-dn>" and filter "<filter>" did not return any results".
+----
+
+Failures of this type most often indicate mis-configured LDAP group configurations
+on the LDAP server. However, if it is certain that the group entries on the LDAP server
+have been configured correctly, these errors can be tolerated and the problematic
+members ignored.
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: This often indicates a mis-configured `usersQuery`. However, in cases where the group contains member entries that are known to be missing, setting `tolerateMemberNotFoundErrors: true` will allow the group sync to continue.

**Comment by stevekuznetsov** [comment]
@liggitt updated, added comment to both sections explicitly point out that tolerate === ignore members

**Comment by liggitt** [comment]
LGTM

File: install_config/syncing_groups_with_ldap.adoc [hunk 9]
```diff
@@ -433,6 +437,156 @@ users:
 <1> The name of the Group as specified by the user-defined name mapping.
 ====
 
+[[rfc2307-with-error-tolerances]]
+=== RFC 2307 with User-Defined Error Tolerances
+
+By default, if groups being synced contain members whose entries that are outside of the
```
Comment (adellape) [inline install_config/syncing_groups_with_ldap.adoc]: s/that//

File: install_config/syncing_groups_with_ldap.adoc [hunk 10]
```diff
@@ -433,6 +437,156 @@ users:
...
+  objectClass: groupOfNames
+  cn: admins
+  owner: cn=admin,dc=example,dc=com
+  description: System Administrators
+  member: cn=Jane,ou=users,dc=example,dc=com
+  member: cn=Jim,ou=users,dc=example,dc=com
+  member: cn=INVALID,ou=users,dc=example,dc=com <1>
+  member: cn=Jim,ou=OUTOFSCOPE,dc=example,dc=com <2>
+----
+<1> This member doesn't exist on the LDAP server.
```
Comment (adellape) [inline install_config/syncing_groups_with_ldap.adoc]: s/doesn't/does not/

File: install_config/syncing_groups_with_ldap.adoc [hunk 11]
```diff
@@ -433,6 +437,156 @@ users:
...
+<2> This member may exist, but is not under the base DN in the
+user query for the sync job.
+====
+
+In order to tolerate errors, the following additions to your sync configuration
```
Comment (adellape) [inline install_config/syncing_groups_with_ldap.adoc]: Maybe:

s/In order to tolerate errors,/In order to tolerate the errors in the above example,/

File: install_config/syncing_groups_with_ldap.adoc (hunk 6 above)
Comment (adellape) [inline install_config/syncing_groups_with_ldap.adoc]: @stevekuznetsov Is this still TODO?

**Comment by adellape** [comment]
@stevekuznetsov Few comments inline.

File: install_config/syncing_groups_with_ldap.adoc (hunk 6 above)
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: No.

**Comment by stevekuznetsov** [comment]
@adellape comments addressed

**Comment by stevekuznetsov** [comment]
@adellape is this good to go, or does the pull need more from me?

**Comment by adellape** [comment]
@stevekuznetsov Sorry, changes LGTM.

**Comment by stevekuznetsov** [comment]
@adellape cheers, thanks

**Comment by adellape** [comment]
[rev_history]
|link:../install_config/syncing_groups_with_ldap.html[Syncing Groups With LDAP]
|Added a link:../install_config/syncing_groups_with_ldap.html#rfc2307-with-error-tolerances[RFC 2307 with User-Defined Error Tolerances] section.
%

**Pull request merged**